### ISSUE TITLE
Explicit constructor name vs .init   

### ISSUE THREAD

**Issue by rudsberg** [sambhav-fueled/swift-style-guide#14]
I suggest we add a style guide for when to use `.init` and when to write out the explicit constructor name. One suggestion is to opt for explicit names _where it makes sense_, for instance when initializing attributes:

```
// Preferred
let activeSortOption = CurrentValueSubject<SortOption, Never>()
// Not preferred
let activeSortOption: CurrentValueSubject<SortOption, Never> = .init()
```

But in other cases using `init` does make sense, for instance:

```
// Preferred 
func sortViewModelFactory() -> SortSelectionViewController<SortOption>.ViewModel {
	.init(selected: sortOption, enabledSortOptions: SortOption.allCases)
}

// Not preferred
func sortViewModelFactory() -> SortSelectionViewController<SortOption>.ViewModel {
	SortSelectionViewController<SortOption>.ViewModel(
		selected: sortOption,
		enabledSortOptions: SortOption.allCases
	)
}
```

The first is better because it removes code duplication and is more succinct while still preserving readability and intent. 

I suggest a pragmatic approach where the shorthand `init` is used where it makes sense and then opt for explicit constructors in cases like the first one. And like always with choosing a coding style, deciding what we think is best here will reduce communication overhead and lead to more consistent code across Fueled's code bases. 

Me and @heymansmile had a discussion about this on [this PR](https://github.com/Fueled/dime-ios/pull/418#discussion_r942352249). 

Let's have a discussion here to nail down a solid strategy 😎

**Comment by heymansmile** [comment]
Regarding second example `Preferred` case. This should be formalized more precisely because the function can have a complex body:
```
// Preferred 
func sortViewModelFactory() -> SortSelectionViewController<SortOption>.ViewModel {
	...
	...
	...
	...
	...
	return .init(selected: sortOption, enabledSortOptions: SortOption.allCases)
}
```
So it's not obvious what is returned here until you take a look at the signature.

**Comment by rudsberg** [comment]
@heymansmile Agreed – more formalized is better. As the body gets more complex and thus the `.init` getting further away from the function signature, that indeed increases the rationale for choosing the explicit constructor. Still, for a short to medium function body, the type is easily inferred by the reader. Also considering that one usually starts (at least me!) reading the function signature before diving into the function body.

**Comment by harishsaini** [comment]
In my opinion, the original suggestions makes sense even if the body is complex we know what function is expected to return and `.init` saves a lot of unnecessary code 👍  

// Preferred
let activeSortOption = CurrentValueSubject<SortOption, Never>()
// Not preferred
let activeSortOption: CurrentValueSubject<SortOption, Never> = .init()
But in other cases using init does make sense, for instance:

// Preferred 
func sortViewModelFactory() -> SortSelectionViewController<SortOption>.ViewModel {
	.init(selected: sortOption, enabledSortOptions: SortOption.allCases)
}

// Not preferred
func sortViewModelFactory() -> SortSelectionViewController<SortOption>.ViewModel {
	SortSelectionViewController<SortOption>.ViewModel(
		selected: sortOption,
		enabledSortOptions: SortOption.allCases
	)
}

**Comment by heymansmile** [comment]
Just leaving a bit of historical context here, because we in Fueled often reference the Google Swift Style guide:
https://google.github.io/swift/#initializers-1
which says: _Explicitly calling .init(...) is allowed only when the receiver of the call is a metatype variable._

And anyway independently to the final decision about explicit calling of .init(...) I wouldn't call the way of using initializer with the literal type name a `Not preferred`, this way should be always allowed.

**Comment by heymansmile** [comment]
As a variant we may say that Explicit call of .init(...) is **allowed** (not preferred) in cases where in a function or a calculated property the `return` can be omitted:
```
func foo() -> Stuff {
	.init()
}
```

**Comment by rudsberg** [comment]
> Just leaving a bit of historical context here, because we in Fueled often reference the Google Swift Style guide: https://google.github.io/swift/#initializers-1 which says: _Explicitly calling .init(...) is allowed only when the receiver of the call is a metatype variable._
> 

To me it looks like they don't talk about the same thing as we are. They discuss using `Type.init(...)` rather than only `.init(...)`.

> As a variant we may say that Explicit call of .init(...) is **allowed** (not preferred) in cases where in a function or a calculated property the `return` can be omitted:
> 
> ```
> func foo() -> Stuff {
> 	.init()
> }
> ```

This may be too restraining since it's only applicable for functions and computed properties that have a single expression.
